### PR TITLE
CDH: add get_secret support for Aliyun KMS

### PR DESCRIPTION
**Pull request by 1570005763**
Apply `get_secret()` function for `Getter` trait.

And two methods of accessing Aliyun KMS is offered. One is through ClientKey, the same as encrypt/decrypt function, the other is through EcsRamRole, which can only be used on Aliyun ECS. More information is updated to the doc file.

Again it is hard to test code on CI. I can only test them on my machine, this is a screenshot of the test results.
image

### CODE REVIEW

**Comment by 1570005763** [comment]
The basic build check failure is irrelative to this PR. I'll open another PR to fix it.

**Comment by 1570005763** [comment]
The basic build check failure is fixed in #424.

File: confidential-data-hub/kms/src/plugins/aliyun/credential.rs [hunk 1]
```diff
@@ -113,3 +147,71 @@ impl Credential {
...
+                let mut signer = Signer::new(MessageDigest::sha1(), &pkey)?;
+                signer.update(str_to_sign.as_bytes())?;
+                let signature = signer.sign_to_vec()?;
+                Ok(STANDARD.encode(signature))
+            },
+            _=> Err(anyhow!("Require 'Credential::CredentialClientKey' in 'sign_with_credential_ecs_ram_role()'!"))
+        }
+    }
+
+    pub(crate) fn urlencode_openapi(&self, s: &String) -> String {
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/aliyun/credential.rs:211]:
```suggestion
pub(crate) fn urlencode_openapi(&self, s: &str) -> String {
```

When use a reference to a String, `&str` is preferred. 

BTW, is this function actually used?

**Review by Xynnn007 — COMMENTED** [review]
Thanks @1570005763 for this patch. Here are some comments

File: confidential-data-hub/kms/src/plugins/aliyun/credential.rs (hunk 1 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/credential.rs:211]: Yeah, it's used in both `build_params()` and `do_request()` functions.

File: confidential-data-hub/kms/src/plugins/aliyun/client.rs [hunk 2]
```diff
@@ -18,12 +18,12 @@ use serde_json::Value;
 use sha2::{Digest, Sha256};
 use tokio::fs;
 
-use crate::plugins::aliyun::client::dkms_api::{DecryptRequest, EncryptRequest};
+// use crate::plugins::aliyun::client::dkms_api;
```
[the remark this answers is not in the document]
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/client.rs]: deleted

File: confidential-data-hub/kms/src/plugins/aliyun/client.rs [hunk 3]
```diff
@@ -355,4 +397,31 @@ mod tests {
...
+
+    #[ignore]
+    #[tokio::test]
+    async fn get_secret_with_client_key() {
+        let secret_name = "test_secret";
+        let provider_settings = json!({
+            "client_key_id": "KAAP.2bc431c0-416f-4ccb-8638-e71245fd60c1",
+            "kms_instance_id": "kst-bjj65794c36xm0v9aeujs",
+        });
+        // init encrypter at user side
```
[the remark this answers is not in the document]
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/client.rs]: Modified.

File: confidential-data-hub/kms/src/plugins/aliyun/client.rs [hunk 4]
```diff
@@ -355,4 +397,31 @@ mod tests {
 
         assert_eq!(decrypted, plaintext);
     }
+
+    #[ignore]
+    #[tokio::test]
+    async fn get_secret_with_client_key() {
+        let secret_name = "test_secret";
```
[the remark this answers is not in the document]
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/client.rs:729]: Sure, `rtest` looks better.

File: confidential-data-hub/kms/src/plugins/aliyun/config.rs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+use anyhow::*;
+// use base64::{engine::general_purpose::STANDARD, Engine};
+// use log::debug;
+// use openssl::{
+//     pkey::{PKey, Private},
+//     sign::Signer,
+// };
+// use p12::{CertBag, ContentInfo, MacData, SafeBagKind, PFX};
+// use serde::Deserialize;
+// use yasna::ASN1Result;
```
[the remark this answers is not in the document]
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/config.rs]: Deleted.

File: confidential-data-hub/kms/src/plugins/aliyun/credential.rs [hunk 6]
```diff
@@ -113,3 +147,71 @@ impl Credential {
...
+                    return Err(anyhow!(
+                        "Request session_credential failed with status: {}",
+                        response.status()
+                    ));
+                };
+                let body_json: Value = serde_json::from_str(&body)?;
+
+                let session_ak = body_json["AccessKeyId"].as_str().unwrap().to_string();
+                let session_sk = body_json["AccessKeySecret"].as_str().unwrap().to_string();
+                let token = body_json["SecurityToken"].as_str().unwrap().to_string();
```
[the remark this answers is not in the document]
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/credential.rs]: Helpful suggestion! Has been modified.

File: confidential-data-hub/kms/src/plugins/aliyun/credential.rs (hunk 1 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/credential.rs]: Modified.

File: confidential-data-hub/kms/src/plugins/aliyun/credential.rs [hunk 7]
```diff
@@ -35,25 +38,30 @@ struct Password {
...
+        let credential = CredentialClientKey {
             client_key_id: ck.key_id.clone(),
             private_key,
         };
 
         Ok(credential)
     }
 
-    pub(crate) fn generate_bear_auth(&self, str_to_sign: &str) -> Result<String> {
+    pub(crate) fn sign(&self, str_to_sign: &str) -> Result<String> {
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/aliyun/credential.rs:64]: This function actually generates bear auth string. Do we need to change the name?

**Review by Xynnn007 — CHANGES_REQUESTED** [review]

File: confidential-data-hub/kms/src/plugins/aliyun/credential.rs (hunk 7 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/credential.rs:64]: I was originally hoping to unify the function names for obtaining signatures in clientKey access and ecsRamRole access. It seems that the `sign()` might be confusing. What do you think about changing it to `get_signature()`?

**Comment by 1570005763** [comment]
I have reorganized the code structure to make it easier to understand. Please take another look. @Xynnn007

File: confidential-data-hub/kms/src/plugins/aliyun/client/client_key_client/config.rs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+#[derive(Clone, Debug)]
+pub(crate) struct ConfigClientKey {
+    pub kms_instance_id: String,
+    pub endpoint: String,
+}
+
+// implement ConfigClientKey related function
+impl ConfigClientKey {
+    pub(crate) fn new(kms_instance_id: &str, endpoint: &str) -> Result<Self> {
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/aliyun/client/client_key_client/config.rs:18]: If we do not return any `Error`, usually we do not use `Result<Self>` but directly `Self`

File: confidential-data-hub/kms/src/plugins/aliyun/client/client_key_client/config.rs [hunk 9]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) 2023 Alibaba Cloud
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/aliyun/client/client_key_client/config.rs:1]: 2023 -> 2024

File: confidential-data-hub/kms/src/plugins/aliyun/client/ecs_ram_role_client/config.rs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+#[derive(Clone, Debug)]
+pub(crate) struct ConfigEcsRamRole {
+    pub region_id: String,
+    pub endpoint: String,
+    pub _vpc: String,
+}
+
+// implement ConfigEcsRamRole related function
+impl ConfigEcsRamRole {
+    pub(crate) fn new(region_id: &str, endpoint: &str, vpc: &str) -> Result<Self> {
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/aliyun/client/ecs_ram_role_client/config.rs:19]: The same issue upon `Result<_>` usage

File: confidential-data-hub/kms/src/plugins/aliyun/client/ecs_ram_role_client/mod.rs [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+            ));
+        };
+
+        Self::new(ecs_ram_role_name, region_id)
+    }
+
+    /// Export the [`ProviderSettings`] of the current client. This function is to be used
+    /// in the encryptor side. The [`ProviderSettings`] will be used to initial a client
+    /// in the decryptor side.
+    pub fn export_provider_settings(&self) -> Result<ProviderSettings> {
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/aliyun/client/ecs_ram_role_client/mod.rs:106]: The same issue upon `Result<_>` usage.

File: confidential-data-hub/kms/src/plugins/aliyun/client/mod.rs [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+                        Error::AliyunKmsError(format!(
+                            "build EcsRamRoleClient with `from_provider_settings()` failed: {e}"
+                        ))
+                    })?,
+            }
+        } else {
+            return Err(Error::AliyunKmsError(
+                "client type invalid or not exist.".to_string(),
+            ));
+        };
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/aliyun/client/mod.rs:104]: Could you do some refactoring like the following to make the logic clearer
```suggestion
        let Some(client_type) = provider_settings.get("client_type") else {
                     return Err(Error::AliyunKmsError(
                "client type invalid or not exist.".to_string(),
            ));
        };
        
        let inner_client = match client_type {
            "client_key" => ...
            "ecs_ram_role" => ...
        }
```

**Review by Xynnn007 — COMMENTED** [review]
Let's make it better

File: confidential-data-hub/kms/src/plugins/aliyun/client/mod.rs (hunk 12 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/client/mod.rs:104]: Great! It does look clearer.

File: confidential-data-hub/kms/src/plugins/aliyun/client/ecs_ram_role_client/mod.rs (hunk 11 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/client/ecs_ram_role_client/mod.rs]: `Result<_>` removed.

File: confidential-data-hub/kms/src/plugins/aliyun/client/ecs_ram_role_client/config.rs (hunk 10 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/client/ecs_ram_role_client/config.rs]: `Result<_>` removed.

File: confidential-data-hub/kms/src/plugins/aliyun/client/client_key_client/config.rs (hunk 9 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/client/client_key_client/config.rs]: updated.

File: confidential-data-hub/kms/src/plugins/aliyun/client/client_key_client/config.rs (hunk 8 above)
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/client/client_key_client/config.rs]: `Result<_>` removed.

File: confidential-data-hub/kms/Cargo.toml [hunk 13]
```diff
@@ -21,17 +21,19 @@ log.workspace = true
 openssl = { workspace = true, optional = true }
 p12 = { version = "0.6.3", optional = true }
 prost = { workspace = true, optional = true }
+rand.workspace = true
```
[the remark this answers is not in the document]
Comment (1570005763) [reply confidential-data-hub/kms/Cargo.toml]: modified.

File: confidential-data-hub/kms/Cargo.toml [hunk 14]
```diff
@@ -21,17 +21,19 @@ log.workspace = true
...
 serde.workspace = true
 serde_json.workspace = true
 sev = { path = "../../attestation-agent/deps/sev", optional = true }
 strum.workspace = true
-reqwest = { version = "0.11", optional = true }
 thiserror.workspace = true
 tokio = { workspace = true, features = ["fs"] }
 toml.workspace = true
 tonic = { workspace = true, optional = true }
+url.workspace = true
```
[the remark this answers is not in the document]
Comment (1570005763) [reply confidential-data-hub/kms/Cargo.toml]: modified.

File: .gitignore [hunk 15]
```diff
@@ -3,4 +3,5 @@
 .DS_Store
 
 image-rs/scripts/attestation-agent
-shell.nix
+confidential-data-hub/kms/src/plugins/aliyun/example_credential/
```
[the remark this answers is not in the document]
Comment (1570005763) [reply .gitignore:6]: Yeah. This directory is used frequently when testing code locally. And it contains secret info to access Aliyun KMS instance. So I add it to `.gitignore`.

File: confidential-data-hub/kms/src/plugins/aliyun/client.rs [hunk 16]
```diff
@@ -223,99 +339,347 @@ impl Decrypter for AliyunKmsClient {
...
+                            "{}={}",
+                            credential.urlencode_openapi(k),
+                            credential.urlencode_openapi(v),
+                        )
+                    })
+                    .collect::<Vec<String>>()
+                    .join("&");
+                let urlencoded_canonicalized_params: String =
+                    credential.urlencode_openapi(&canonicalized_params);
+                let string_to_sign = format!("POST&%2F&{}", urlencoded_canonicalized_params,);
```
[the remark this answers is not in the document]
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/client.rs]: deleted.

File: confidential-data-hub/kms/src/plugins/aliyun/client.rs [hunk 17]
```diff
@@ -64,75 +96,159 @@ impl AliyunKmsClient {
...
-        )
+            let ecs_ram_role_json: serde_json::Value = serde_json::from_str(&ecs_ram_role_str)
+                .map_err(|e| {
+                    Error::AliyunKmsError(format!(
+                        "read ecs_ram_role with `serde_json::from_str()` failed: {e}"
+                    ))
+                })?;
+
+            let ecs_ram_role_name =
+                if let Some(ecs_ram_role_name) = ecs_ram_role_json["ecs_ram_role_name"].as_str() {
```
[the remark this answers is not in the document]
Comment (1570005763) [reply confidential-data-hub/kms/src/plugins/aliyun/client.rs]: modified.

File: confidential-data-hub/kms/src/plugins/aliyun/client/ecs_ram_role_client/mod.rs [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+            config,
+            http_client,
+        })
+    }
+
+    /// This new function is used by a in-pod client. The side-effect is to read the
+    /// [`ALIYUN_IN_GUEST_DEFAULT_KEY_PATH`] which is the by default path where the credential
+    /// to access kms is saved.
+    pub async fn from_provider_settings(_provider_settings: &ProviderSettings) -> Result<Self> {
+        let ecs_ram_role_path = format!("{ALIYUN_IN_GUEST_DEFAULT_KEY_PATH}/ecsRamRole.json");
```
Comment (Xynnn007) [inline confidential-data-hub/kms/src/plugins/aliyun/client/ecs_ram_role_client/mod.rs:61]: We can use `concatcp!` to declare this as a const.

**Review by Xynnn007 — APPROVED** [review]
LGTM.

**Pull request merged**